libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/AppLockView/SFSDKAppLockViewController.m

```diff
@@ -0,0 +1,226 @@
+        [self promptBiometricEnrollment:passcode];
+    } else {
+        SFSecurityLockoutAction action = isUpdateMode?SFSecurityLockoutActionPasscodeChanged:SFSecurityLockoutActionPasscodeCreated;
+        [self unlock:passcode lockAction:action];
+    }
+}
+
+#pragma mark - SFSDKPasscodeVerifyDelegate
+
+- (void)passcodeVerified:(NSString *)passcode
+{
+    if ([self canPromptBiometricEnrollment]) {
+        [self promptBiometricEnrollment:passcode];
+    } else {
+        SFSecurityLockoutAction action = SFSecurityLockoutActionPasscodeVerified;
+        [self unlock:passcode lockAction:action];
+    }
+}
+
+- (void)passcodeFailed
+{
+    [[SFPasscodeManager sharedManager] resetPasscode];
+    [SFSecurityLockout unlock:NO action:SFSecurityLockoutActionNone passcodeConfig:self.configData];
+}
+
+#pragma mark - SFSDKBiometricViewDelegate
+
+- (void)biometricUnlockSucceeded:(NSString *)currentPasscode verificationMode:(BOOL)isVerificationMode
+{
+    [SFSecurityLockout setBiometricUnlockEnabled:YES];
+    
+    if (!isVerificationMode && !currentPasscode) {
+        [self dismissStandaloneBiometricSetup];
+    } else {
+        SFSecurityLockoutAction action = isVerificationMode?SFSecurityLockoutActionPasscodeChanged:SFSecurityLockoutActionPasscodeCreated;
+        [self.navigationController popViewControllerAnimated:NO];
+        [self unlock:currentPasscode lockAction:action];
+    }
+}
+
+- (void)biometricUnlockFailed:(NSString *)currentPasscode verificationMode:(BOOL)isVerificationMode
+{
+    if (isVerificationMode) {
+        [self.navigationController popViewControllerAnimated:NO];
+        SFSDKPasscodeVerifyController *pvc = [[SFSDKPasscodeVerifyController alloc] initWithViewConfig:self.viewConfig];
+        pvc.verifyDelegate = self;
+        [self pushViewController:pvc animated:NO];
+    } else {
+        // No passcode means Biometric Setup screen was launched outside of Lock Flow
+        if (currentPasscode) {
+            [SFSecurityLockout setUserDeclinedBiometricUnlock:YES];
+            SFSecurityLockoutAction action = isVerificationMode?SFSecurityLockoutActionPasscodeVerified:SFSecurityLockoutActionPasscodeCreated;
+            [self unlock:currentPasscode lockAction:action];
+        } else {
+            [self dismissStandaloneBiometricSetup];
+        }
+    }
+}
+
+- (void)dismissStandaloneBiometricSetup
+{
+    [SFSecurityLockout setupTimer];
+    [[[SFSDKWindowManager sharedManager] passcodeWindow].viewController dismissViewControllerAnimated:NO completion:^{
+        [[SFSDKWindowManager sharedManager].passcodeWindow dismissWindowAnimated:NO withCompletion:^{}];
+    }];
+}
+
+-(UIViewController *)controllerFromMode:(SFAppLockControllerMode) mode configData:(SFAppLockConfigurationData)configData andViewConfig:(SFSDKAppLockViewConfig *)viewConfig
+{
+    UIViewController *currentViewController = nil;
+    
+    if (mode == SFAppLockControllerModeEnableBiometric || mode == SFAppLockControllerModeVerifyBiometric) {
```

Comment:
I thought the same thing since you can just fall through for duplicate cases, but I decided against it since you can't declare variables inside a case block.  Is declaring `SFSDKBiometricViewController *bvc`,  `SFSDKPasscodeCreateController *pvc` and `SFSDKPasscodeVerifyController *pvc` first and initializing inside the switch is probably still cleaner?

Reply:
I think the switch would still be more readable - also if we add more modes, compiler could tell us that they are not yet handled.